### PR TITLE
Nav bar from topic api

### PR DESCRIPTION
**Pull request by liamjlewis**
### What

Navigation bar data is taken from dp-topic-api and sent through to dp-renderer. It is cached using the dp-cache library.

### How to review

No special instructions.

### Who can review

Anyone.

### CODE REVIEW

File: cache/mock.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		return nil, err
+	}
+
+	mockNavigationData := &models.Navigation{
+		Description: "this is a test description",
+	}
+
+	for _, v := range langs {
+		navigationlangKey := testNavigationCache.GetCachingKeyForNavigationLanguage(v)
+
```
Comment (justinpjose) [inline cache/mock.go:37]: This empty space can be removed

File: cache/navigation.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	key := nc.GetCachingKeyForNavigationLanguage(lang)
+	navigationCacheInterface, ok := nc.Get(key)
+	if !ok {
+		err := fmt.Errorf("cached navigation data with key %s not found", key)
+		log.Error(ctx, "failed to get cached navigation data", err)
+		return &models.Navigation{}, err
+	}
+
+	navigationCache, ok := navigationCacheInterface.(*models.Navigation)
+
```
Comment (justinpjose) [inline cache/navigation.go:61]: Can remove empty space here as well

File: handlers/dataset.go [hunk 3]
```diff
@@ -64,15 +67,24 @@ func datasetPage(w http.ResponseWriter, req *http.Request, zc ZebedeeClient, ren
...
+	}
+
 	basePage := rend.NewBasePageModel()
-	m := mapper.CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versions, lang, homepageContent.ServiceMessage, homepageContent.EmergencyBanner)
+	m := mapper.CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versions, lang, homepageContent.ServiceMessage, homepageContent.EmergencyBanner, navigationCache)
 
 	rend.BuildPage(w, m, "dataset")
 }
 
 // DataSet will load a legacy dataset page
```
Comment (justinpjose) [inline handlers/dataset.go:85]:
```suggestion
// DatasetPage will load a legacy dataset page
```

File: handlers/dataset_test.go [hunk 4]
```diff
@@ -89,7 +92,15 @@ func TestDatasetHandlers(t *testing.T) {
 
 				Convey("When the dataset page is rendered", func() {
 					w, req := generateRecorderAndRequest()
-					DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+					langs := []string{}
```
Comment (justinpjose) [inline handlers/dataset_test.go:96]:
```suggestion
					langs := []string{"en","cy"}
```

Comment (justinpjose) [reply handlers/dataset_test.go:96]: **COMMENT 1**
Then you are remove the below two lines of appending

File: handlers/dataset_test.go [hunk 5]
```diff
@@ -106,7 +117,15 @@ func TestDatasetHandlers(t *testing.T) {
 				mockFilesAPIClient.EXPECT().GetFile(ctx, expectedDownloadFilename, userAuthTokenDatasets).Return(files.FileMetaData{}, errors.New("files api broken"))
 
 				w, req := generateRecorderAndRequest()
-				DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+				langs := []string{}
+				langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/dataset_test.go:122]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 6]
```diff
@@ -177,7 +203,15 @@ func TestDatasetHandlers(t *testing.T) {
 
 				Convey("When the dataset page is rendered", func() {
 					w, req := generateRecorderAndRequest()
-					DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+					langs := []string{}
+					langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/dataset_test.go:208]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 7]
```diff
@@ -191,7 +225,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:230]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 8]
```diff
@@ -205,7 +247,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:252]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 9]
```diff
@@ -219,7 +269,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:274]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 10]
```diff
@@ -235,7 +293,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:298]: Same as **COMMENT 1**

File: handlers/dataset_test.go [hunk 11]
```diff
@@ -255,7 +321,15 @@ func TestDatasetHandlers(t *testing.T) {
 			req.URL.Path = path + "/data"
 
 			mockZebedeeClient.EXPECT().Get(ctx, userAuthTokenDatasets, zebedeePath).Return(expectedBody, nil)
-			DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/dataset_test.go:326]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 12]
```diff
@@ -37,7 +40,14 @@ func TestLegacyLanding(t *testing.T) {
 			req, _ := http.NewRequest(http.MethodGet, path+"/data", nil)
 			req.AddCookie(&http.Cookie{Name: "access_token", Value: "12345"})
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil)(w, req)
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:44]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 13]
```diff
@@ -60,7 +70,14 @@ func TestLegacyLanding(t *testing.T) {
 			w := httptest.NewRecorder()
 			req, _ := http.NewRequest(http.MethodGet, "/somelegacypage", nil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, mockRend).ServeHTTP(w, req)
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:74]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 14]
```diff
@@ -72,7 +89,14 @@ func TestLegacyLanding(t *testing.T) {
 			w := httptest.NewRecorder()
 			req, _ := http.NewRequest(http.MethodGet, "/somelegacypage", nil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil).ServeHTTP(w, req)
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:93]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 15]
```diff
@@ -86,7 +110,14 @@ func TestLegacyLanding(t *testing.T) {
 			req, err := http.NewRequest("GET", "/somelegacypage", nil)
 			So(err, ShouldBeNil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil).ServeHTTP(w, req)
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:114]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 16]
```diff
@@ -149,7 +180,13 @@ func TestHandlersFilesAPI(t *testing.T) {
 			req, _ := http.NewRequest("GET", legacyURL, nil)
 			req.Header.Set(authHeaderKey, expectedAuthToken)
 
-			handler := LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, mockRend)
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:184]: Same as **COMMENT 1**

File: handlers/legacy_landing_test.go [hunk 17]
```diff
@@ -195,7 +232,13 @@ func TestHandlersFilesAPI(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:236]: Same as **COMMENT 1**

File: mapper/dataset_test.go [hunk 18]
```diff
@@ -244,9 +247,21 @@ func TestCreateDatasetPage(t *testing.T) {
 		serviceMessage := getTestServiceMessage()
 		emergencyBanner := getTestEmergencyBanner()
 
+		// get cached navigation data
+		langs := []string{}
+		langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/dataset_test.go:252]: Same as **COMMENT 1**

File: mapper/dataset_test.go [hunk 19]
```diff
@@ -321,7 +336,20 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
 		}
 
 		Convey("When CreateDatasetPage is called", func() {
-			result := CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versions, lang, serviceMessage, emergencyBanner)
+
+			// get cached navigation data
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/dataset_test.go:342]: Same as COMMENT 1

File: mapper/dataset_test.go [hunk 20]
```diff
@@ -356,7 +384,19 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
 		}
 
 		Convey("When CreateDatasetPage is called", func() {
-			result := CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versionedDatasets, lang, serviceMessage, emergencyBanner)
+			// get cached navigation data
+			langs := []string{}
+			langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/dataset_test.go:389]: Same as COMMENT 1

File: mapper/dataset_test.go [hunk 21]
```diff
@@ -412,7 +452,19 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
...
```
Comment (justinpjose) [inline mapper/dataset_test.go:457]: Same as COMMENT 1

File: mapper/legacy_test.go [hunk 22]
```diff
@@ -24,7 +26,19 @@ func TestUnitMapperLegacy(t *testing.T) {
 		serviceMessage := getTestServiceMessage()
 		emergencyBanner := getTestEmergencyBanner()
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		langs := []string{}
+		langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/legacy_test.go:31]: Same as COMMENT 1

File: mapper/legacy_test.go [hunk 23]
```diff
@@ -87,7 +101,19 @@ func TestUnitMapperLegacy(t *testing.T) {
 
 		ds := zebedeeOnlyTestDatasets(expectedDatasetURI, expectedFilename, expectedSupplementaryTitle)
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		langs := []string{}
+		langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/legacy_test.go:106]: Same as COMMENT 1

File: mapper/legacy_test.go [hunk 24]
```diff
@@ -116,7 +142,19 @@ func TestUnitMapperLegacy(t *testing.T) {
 		expectedSupplementaryFilepath := "data/collection-id/new-file.xlsx"
 		ds := staticFilesOnlyTestDatasets(expectedDownloadFilepath, expectedSupplementaryTitle, expectedSupplementaryFilepath)
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		langs := []string{}
+		langs = append(langs, "en")
```
Comment (justinpjose) [inline mapper/legacy_test.go:147]: Same as COMMENT 1

**Review by justinpjose — COMMENTED** [review]
Two tiny comments
All the other comments are of the same

File: handlers/dataset_test.go [hunk 25]
```diff
@@ -89,7 +92,15 @@ func TestDatasetHandlers(t *testing.T) {
 
 				Convey("When the dataset page is rendered", func() {
 					w, req := generateRecorderAndRequest()
-					DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+					cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/dataset_test.go:96]: **COMMENT 1**
You don't need to do this as you already have got `cfg` in this test at line 53 which contains `SupportedLanguages`. You can remove lines 96 to 98

File: handlers/dataset_test.go [hunk 26]
```diff
@@ -89,7 +92,15 @@ func TestDatasetHandlers(t *testing.T) {
...
+					So(err, ShouldBeNil)
+
+					ctxOther := context.Background()
```
Comment (justinpjose) [inline handlers/dataset_test.go:99]: **COMMENT 2**
You can just use the `ctx` already defined in the test at line 43 rather creating another context (`ctxOther`)

File: handlers/dataset_test.go [hunk 27]
```diff
@@ -106,7 +117,15 @@ func TestDatasetHandlers(t *testing.T) {
 				mockFilesAPIClient.EXPECT().GetFile(ctx, expectedDownloadFilename, userAuthTokenDatasets).Return(files.FileMetaData{}, errors.New("files api broken"))
 
 				w, req := generateRecorderAndRequest()
-				DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+				cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/dataset_test.go:121]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 28]
```diff
@@ -155,7 +174,14 @@ func TestDatasetHandlers(t *testing.T) {
 
 				Convey("When the dataset page is rendered", func() {
 					w, req := generateRecorderAndRequest()
-					DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+					cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/dataset_test.go:178]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 29]
```diff
@@ -177,7 +203,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:207]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 30]
```diff
@@ -191,7 +225,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:229]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 31]
```diff
@@ -205,7 +247,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:251]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 32]
```diff
@@ -219,7 +269,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:273]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 33]
```diff
@@ -235,7 +293,15 @@ func TestDatasetHandlers(t *testing.T) {
...
```
Comment (justinpjose) [inline handlers/dataset_test.go:297]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/dataset_test.go [hunk 34]
```diff
@@ -255,7 +321,15 @@ func TestDatasetHandlers(t *testing.T) {
 			req.URL.Path = path + "/data"
 
 			mockZebedeeClient.EXPECT().Get(ctx, userAuthTokenDatasets, zebedeePath).Return(expectedBody, nil)
-			DatasetPage(mockZebedeeClient, mockRend, mockFilesAPIClient)(w, req)
+
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/dataset_test.go:325]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/legacy_landing_test.go [hunk 35]
```diff
@@ -37,7 +40,14 @@ func TestLegacyLanding(t *testing.T) {
 			req, _ := http.NewRequest(http.MethodGet, path+"/data", nil)
 			req.AddCookie(&http.Cookie{Name: "access_token", Value: "12345"})
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil)(w, req)
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:43]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/legacy_landing_test.go [hunk 36]
```diff
@@ -60,7 +70,14 @@ func TestLegacyLanding(t *testing.T) {
 			w := httptest.NewRecorder()
 			req, _ := http.NewRequest(http.MethodGet, "/somelegacypage", nil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, mockRend).ServeHTTP(w, req)
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:73]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/legacy_landing_test.go [hunk 37]
```diff
@@ -72,7 +89,14 @@ func TestLegacyLanding(t *testing.T) {
 			w := httptest.NewRecorder()
 			req, _ := http.NewRequest(http.MethodGet, "/somelegacypage", nil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil).ServeHTTP(w, req)
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:92]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/legacy_landing_test.go [hunk 38]
```diff
@@ -86,7 +110,14 @@ func TestLegacyLanding(t *testing.T) {
 			req, err := http.NewRequest("GET", "/somelegacypage", nil)
 			So(err, ShouldBeNil)
 
-			LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, nil).ServeHTTP(w, req)
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:113]: Same as **COMMENT 1** and **COMMENT 2**

File: handlers/legacy_landing_test.go [hunk 39]
```diff
@@ -195,7 +232,13 @@ func TestHandlersFilesAPI(t *testing.T) {
 			req, _ := http.NewRequest("GET", legacyURL, nil)
 			req.Header.Set(authHeaderKey, expectedAuthToken)
 
-			handler := LegacyLanding(mockZebedeeClient, mockDatasetClient, mockFilesAPIClient, mockRend)
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline handlers/legacy_landing_test.go:235]: Same as **COMMENT 1** and **COMMENT 2**

File: mapper/dataset_test.go [hunk 40]
```diff
@@ -244,9 +248,20 @@ func TestCreateDatasetPage(t *testing.T) {
 		serviceMessage := getTestServiceMessage()
 		emergencyBanner := getTestEmergencyBanner()
 
+		// get cached navigation data
+		cfg, err := config.Get()
+		So(err, ShouldBeNil)
+		ctxOther := context.Background()
```
Comment (justinpjose) [inline mapper/dataset_test.go:254]: You can just use the `ctx` defined in line 244

File: mapper/dataset_test.go [hunk 41]
```diff
@@ -244,9 +248,20 @@ func TestCreateDatasetPage(t *testing.T) {
 		serviceMessage := getTestServiceMessage()
 		emergencyBanner := getTestEmergencyBanner()
 
+		// get cached navigation data
+		cfg, err := config.Get()
```
Comment (justinpjose) [inline mapper/dataset_test.go:252]: The `cfg` should generally be initialised in the `Given` level of the test or above that

File: mapper/dataset_test.go [hunk 42]
```diff
@@ -321,7 +336,19 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
 		}
 
 		Convey("When CreateDatasetPage is called", func() {
-			result := CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versions, lang, serviceMessage, emergencyBanner)
+
+			// get cached navigation data
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline mapper/dataset_test.go:341]: Same as **COMMENT 3** and **COMMENT 4**

File: mapper/dataset_test.go [hunk 43]
```diff
@@ -356,7 +383,18 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
 		}
 
 		Convey("When CreateDatasetPage is called", func() {
-			result := CreateDatasetPage(basePage, ctx, req, ds, dlp, bc, versionedDatasets, lang, serviceMessage, emergencyBanner)
+			// get cached navigation data
+			cfg, err := config.Get()
```
Comment (justinpjose) [inline mapper/dataset_test.go:387]: Same as **COMMENT 3** and **COMMENT 4**

File: mapper/dataset_test.go [hunk 44]
```diff
@@ -412,7 +450,18 @@ func TestCreateDatasetPageFileLinks(t *testing.T) {
...
```
Comment (justinpjose) [inline mapper/dataset_test.go:454]: Same as **COMMENT 3** and **COMMENT 4**

File: mapper/legacy_test.go [hunk 45]
```diff
@@ -24,7 +27,18 @@ func TestUnitMapperLegacy(t *testing.T) {
 		serviceMessage := getTestServiceMessage()
 		emergencyBanner := getTestEmergencyBanner()
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		cfg, err := config.Get()
+		So(err, ShouldBeNil)
+		ctxOther := context.Background()
```
Comment (justinpjose) [inline mapper/legacy_test.go:33]: Can just use `ctx` instead of creating `ctxOther`

File: mapper/legacy_test.go [hunk 46]
```diff
@@ -87,7 +101,18 @@ func TestUnitMapperLegacy(t *testing.T) {
 
 		ds := zebedeeOnlyTestDatasets(expectedDatasetURI, expectedFilename, expectedSupplementaryTitle)
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		cfg, err := config.Get()
+		So(err, ShouldBeNil)
+		ctxOther := context.Background()
```
Comment (justinpjose) [inline mapper/legacy_test.go:107]: COMMENT 5

File: mapper/legacy_test.go [hunk 47]
```diff
@@ -116,7 +141,18 @@ func TestUnitMapperLegacy(t *testing.T) {
 		expectedSupplementaryFilepath := "data/collection-id/new-file.xlsx"
 		ds := staticFilesOnlyTestDatasets(expectedDownloadFilepath, expectedSupplementaryTitle, expectedSupplementaryFilepath)
 
-		sdlp := CreateLegacyDatasetLanding(mdl, ctx, req, dlp, bcs, ds, lang, serviceMessage, emergencyBanner)
+		// get cached navigation data
+		cfg, err := config.Get()
+		So(err, ShouldBeNil)
+		ctxOther := context.Background()
```
Comment (justinpjose) [inline mapper/legacy_test.go:147]: COMMENT 5

File: handlers/dataset_test.go (hunk 26 above)
Comment (liamjlewis) [reply handlers/dataset_test.go:99]: It doesn't like the type of ctx, if I use ctx it says:
`cannot use ctx (variable of type gomock.Matcher) as "context".Context value in argument to cache.GetMockCacheList: missing method Deadline`
Any ideas for a fix?

**Review by justinpjose — APPROVED** [review]

**Pull request merged**